android/src/main/java/com/oblador/keychain/KeychainModule.java

```diff
@@ -37,99 +41,108 @@ public String getName() {
+            promise.reject(E_CRYPTO_FAILED, e);
         }
     }
 
     @ReactMethod
     public void getGenericPasswordForOptions(String service, Promise promise) {
-        service = service == null ? EMPTY_STRING : service;
-
-        String username = prefs.getString(service + ":u", "user_not_found");
-        String password = prefs.getString(service + ":p", "pass_not_found");
-        if (username.equals("user_not_found") || password.equals("pass_not_found")) {
-            Log.e(KEYCHAIN_MODULE, "no keychain entry found for service: " + service);
-            promise.resolve(false);
-            return;
-        }
-
-        byte[] recuser = Base64.decode(username, Base64.DEFAULT);
-        byte[] recpass = Base64.decode(password, Base64.DEFAULT);
-
-        Entity userentity = Entity.create(KEYCHAIN_DATA + ":" + service + "user");
-        Entity pwentity = Entity.create(KEYCHAIN_DATA + ":" + service + "pass");
-
         try {
-            byte[] decryptedUsername = crypto.decrypt(recuser, userentity);
-            byte[] decryptedPass = crypto.decrypt(recpass, pwentity);
+            CipherStorage currentCipherStorage = getCipherStorageForCurrentAPILevel();
+            if (currentCipherStorage == null) {
+                throw new CryptoFailedException("Unsupported Android SDK " + Build.VERSION.SDK_INT);
+            }
+
+            final DecryptionResult decryptionResult;
+            ResultSet resultSet = prefsStorage.getEncryptedEntry(service);
+            if (resultSet == null) {
+                Log.e(KEYCHAIN_MODULE, "No entry found for service: " + service);
+                promise.resolve(false);
+                return;
+            }
+
+            if (resultSet.cipherStorageName.equals(currentCipherStorage.getCipherStorageName())) {
+                // The encrypted data is encrypted using the current CipherStorage, so we just decrypt and return
+                decryptionResult = currentCipherStorage.decrypt(service, resultSet.usernameBytes, resultSet.passwordBytes);
+            }
+            else {
+                // The encrypted data is encrypted using an older CipherStorage, so we need to decrypt the data first, then encrypt it using the current CipherStorage, then store it again and return
+                CipherStorage oldCipherStorage = getCipherStorageByName(resultSet.cipherStorageName);
+                // decrypt using the older cipher storage
+                decryptionResult = oldCipherStorage.decrypt(service, resultSet.usernameBytes, resultSet.passwordBytes);
+                // encrypt using the current cipher storage
+                EncryptionResult encryptionResult = currentCipherStorage.encrypt(service, decryptionResult.username, decryptionResult.password);
+                // store the encryption result
+                prefsStorage.storeEncryptedEntry(service, encryptionResult);
+                // clean up the old cipher storage
+                oldCipherStorage.removeKey(service);
```

Comment:
for `CipherStorageFacebookConceal`, this will be a no-op. But we want the conceal values be removed from sharedPreferences.

`removeKey` removes the entry from KeyStore on the new `CipherStorageKeystoreAESCBC` implementation of `CypherStorage` interface but does nothing for the `CipherStorageFacebookConceal` implementation of the `CypherStorage` interface. Then, there's the `resetGenericPasswordForOptions` method which calls `removeKey` and also calls `resetPassword` which removes the entries from sharedPreferences.

I suggest we merge the `removeKey` and `resetPassword` into one method because if I understand it correctly, when we remove a key we want to (1) remove it from the sharedPreferences and(2) from the Keystore (if using the KeyStore implementation) and we want this to happen together in one place rather than having it in multiple places.

In other words, when you call `oldCipherStorage.removeKey(service);` (where oldCipherStorage means conceal) you want this call to remove the entry from sharedPreferences. That leads me to thinking that maybe `PrefsStorage prefsStorage` should not be in `KeychainModule` but in the implementations of the `CipherStorage` interface. (Also since `PrefsStorage` has is a utility class, maybe everything in it could be declared static but that is not a big deal now). I would also allow the calls from `KeychainModule` to be more clean - not having to deal with the details of how things are stored, but having the `CipherStorage` implementations deal with it. Then you would just call something like `currentCipherStorage.encryptAndPersist(service, username, password);`